### ISSUE TITLE
Fully publish and support PioneerAnnotationUtils

### ISSUE THREAD

**Issue by eeverman** [junit-pioneer/junit-pioneer#695]
A thought:  This project's support classes are just as valuable (maybe more so?) than its current set of extensions.

I'm creating custom JUnit extensions for a separate project:  [AndHow](https://github.com/eeverman/andhow) (a configuration utility).
JUnit's [AnnotationSupport](https://github.com/junit-team/junit5/blob/732a5400f80c8f446daa8b43eaa4b41b3da929be/junit-platform-commons/src/main/java/org/junit/platform/commons/support/AnnotationSupport.java#L103) methods don't quite work for extensions trying to find their declaring annotation for configuration (i.e. when the annotation has parameters) - I'm guessing that's why this project wrote its own PioneerAnnotationUtils class.

Presumably, the [PioneerAnnotationUtils](https://github.com/junit-pioneer/junit-pioneer/blob/main/src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java) solves the problems of locating annotations from extensions following JUnit's semantics, which would be ***hugely useful*** for anyone else creating extensions.  Unfortunately, the PioneerAnnotationUtils class warns:
> [DO NOT USE THIS CLASS - IT MAY CHANGE SIGNIFICANTLY IN ANY MINOR UPDATE](https://github.com/junit-pioneer/junit-pioneer/blob/1330e8bd50a5088c1ebe833967d045fcd2d9ee74/src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java#L36).

I propose that all the support features used by JUnit Pioneer be split out into a separate artifact, perhaps ***junit-pioneer-support***, and that support methods be marked with support levels similar to what JUnit does for its support methods, e.g. MAINTAINED, EXPERIMENTAL, etc..  That library might become the basis for many other custom built extensions.  I think this is also in keeping with the description of the project:
> It aims to spin off successful and cohesive portions into sibling projects or back into the JUnit 5 code base.

It's hard to spin off extensions if each one has to re-implement all the support pieces that this project has built up :-)

**Comment by Michael1993** [comment]
Hi @eeverman ! Thank you for the suggestion. We will discuss and get back to you.

JUnit's `AnnotationSupport` is a very robust tool, so I'm kind of wondering what you are trying to do to find it lacking. Could you tell us a bit more about your project?

[We had an issue open a while ago about marking our API with API Guardian like you suggested.](https://github.com/junit-pioneer/junit-pioneer/issues/499)
The decision was that we don't want to do that.

**Comment by eeverman** [comment]
My project, [AndHow](https://www.andhowconfig.org), is a project configuration util.  It lets you declare strongly type configuration parameters anywhere in your code like this:
`static final StrProp DATABASE_URL = StrProp.builder().startsWith("jdbc:").build();`
...and then finds and validates the configured values for them from any source (env vars, jndi, sys props, etc.) at startup.

The JUnit extensions I am working on are specifically for AndHow:  The intention is to let the user specify a properties file to be used for a test like this:
```java
@Test
@UseThisPropertiesFileForAndHow("MyFile.properties")
public void someTest() { }
```

So, bottom line, unless you are using AndHow, you wouldn't need these extensions.

**Comment by nipafx** [comment]
It's great to learn that `PioneerAnnotationUtils` provides value beyond its internal use. 😃 I hesitate to spin it into its own project, though. That class has quite a [tumultuous history](https://github.com/junit-pioneer/junit-pioneer/commits/main/src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java) and we have [ideas](https://github.com/junit-pioneer/junit-pioneer/issues/648#issuecomment-1178022653) that may lead to further, deep changes in the future. I'd rather not lock us into a compatibility commitment.

Of course we could just mark it as experimental, but if what you're say is correct, that it "might become the basis for many other custom built extensions", that wouldn't help because then different extension artifacts may depend on different versions of the new support artifact, which can quickly lead to unresolvable conflicts.

Instead I invite you to just copy-paste the source code. EPL 2.0 is pretty permissive, so there's a good chance that it's compatible with your project's license. If it is, just copy the code you need into your utils. 👍🏾 Please change the package name, though, to avoid split packages.

**Comment by eeverman** [comment]
Thank you, @nipafx, for the info.  I would like to do this:  include some of the source code of `PioneerAnnotationUtils` in [AndHow](https://www.andhowconfig.org/), which is under the Apache 2.0 license (likely w/ some small modifications).

Reading through the [EPL 2.0](https://www.eclipse.org/legal/epl-2.0/), it sounds like I may need to dual license that as a result.  Does anyone on this project have the expertise to comment on those actual requirements?  I'd prefer to not have to complicate the license on my project if I don't have to.

Naturally, the [EPL 2.0 FAQs](https://www.eclipse.org/legal/epl-2.0/faq.php) manages to not directly comment on this situation.

**Comment by Michael1993** [comment]
> Does anyone on this project have the expertise to comment on those actual requirements?

I can safely say, no. 😅

**Comment by Bukama** [comment]
Same for me - not a lawyer 🤷‍♂️

**Comment by nipafx** [comment]
I can't say anything definitive on the matter either. But... couldn't you create a separate one-file project under EPL and then depend on that from your project? That would effectively be the same as what you asked us to do minus the issues I laid out earlier.

Given the age of this issue and our lacking desire to implement it, I will close it, but please feel free to reply or reopen if you think we can help more. Also, we'd be interested to learn what solution you went with in the end. 😃
